### PR TITLE
Upgrade dependency

### PR DESCRIPTION
**Pull request by ArieShout**
This PR contains mainly the following changes:

* Use azure-credentials plugin to manage the Azure login credentials
* Upgrade Azure SDK dependency to version 1.1.0

The code had been tested against Azure and it worked with basic examples. I created this pull request as a milestone that we have migrated the original functionalities to the up-to-date dependency set.

The followed tasks will be:

1. Remove the resource group and DC/OS cluster creating logic
2. Clean up the hard coded names and assumptions on the naming pattern in the code base.
3. Integrate with the logic for K8S deployment.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1,141 +1,145 @@
...
-    <!-- Java Level to use. Java 7 required when using core >= 1.612 -->
-    <java.level>7</java.level>
-    <!-- Jenkins Test Harness version you use to test the plugin. -->
-    <!-- For Jenkins version >= 1.580.1 use JTH 2.x or higher. -->
-    <jenkins-test-harness.version>2.1</jenkins-test-harness.version>
-    <!-- Other properties you may want to use:
-         ~ hpi-plugin.version: The HPI Maven Plugin version used by the plugin..
-         ~ stapler-plugin.version: The Stapler Maven plugin version required by the plugin.
+    <artifactId>azure-acs-plugin</artifactId>
+    <version>0.1</version>
```
Comment (chenkennt) [inline pom.xml]: Add SNAPSHOT?

Comment (chenkennt) [inline pom.xml]: VM Agent and Storage plugin are using three digits version, also follow this pattern to use 0.1.0?

File: pom.xml [hunk 2]
```diff
@@ -1,141 +1,145 @@
...
+        <license>
+            <name>MIT License</name>
+            <url>http://opensource.org/licenses/MIT</url>
+        </license>
+    </licenses>
+
+    <!-- If you want this to appear on the wiki page:
+    <developers>
+      <developer>
+        <id>bhacker</id>
```
Comment (chenkennt) [inline pom.xml]: change the names to us.

File: src/main/java/com/microsoft/jenkins/acs/util/NetworkResourceProviderHelper.java [hunk 3]
```diff
@@ -5,200 +5,162 @@
...
+        }
+
+        if (!ruleFound) {
+            context.logStatus("Load balancer rule for port " + hostPort + " not found.");
+            String ruleName = "JLBRuleHttp" + hostPort;
+            context.logStatus("Creating load balancer rule for port " + hostPort + " with name:" + ruleName);
+
+            foundLoadBalancer.update()
+
+                    .defineHttpProbe("httpProbe")
```
Comment (ArchangelSDY) [inline src/main/java/com/microsoft/jenkins/acs/util/NetworkResourceProviderHelper.java]: TCP probe may have better compatibility?

File: src/main/java/com/microsoft/jenkins/acs/util/NetworkResourceProviderHelper.java [hunk 4]
```diff
@@ -5,200 +5,162 @@
...
+            }
+
+            maxPrio = maxPrio + 10;
+            String ruleName = "Allow_" + hostPort;
+            context.logStatus("Creating Security rule for port " + hostPort + " with name:" + ruleName);
+
+            publicGroup.update()
+                    .defineRule(ruleName)
+                    .allowInbound()
+                    .fromAddress("Internet")
```
Comment (ArchangelSDY) [inline src/main/java/com/microsoft/jenkins/acs/util/NetworkResourceProviderHelper.java]: The method comment says that the parameter should be "an IP address prefix expressed in the CIDR notation" so will `Internet` actually work?

File: src/main/java/com/microsoft/jenkins/acs/ACSDeploymentContext.java [hunk 5]
```diff
@@ -19,200 +20,230 @@
...
+    @Override
+    public Azure getAzureClient() {
+        return this.azureClient;
+    }
+
+    public void configure(TaskListener listener, FilePath workspacePath, AzureCredentials.ServicePrincipal servicePrincipal) throws IOException, InterruptedException, AzureCloudException {
+        this.servicePrincipal = servicePrincipal;
+        this.azureClient = Azure.authenticate(DependencyMigration.buildAzureTokenCredentials(servicePrincipal)).withSubscription(servicePrincipal.getSubscriptionId());
+
+        this.localMarathonConfigFile = File.createTempFile("marathon-", ".json", new File(System.getProperty("java.io.tmpdir")));
```
Comment (ArieShout) [inline src/main/java/com/microsoft/jenkins/acs/ACSDeploymentContext.java]: The job is configured on the master node. The configuration file path is configured under the master environment. When the job is running, I assume the slave node needs to read that from a network location, which may need to copy to local in order to feed to the sftp.

File: src/main/java/com/microsoft/jenkins/acs/commands/ValidateContainerCommand.java [hunk 6]
```diff
@@ -5,52 +5,48 @@
...
+    @Override
+    public void execute(ValidateContainerCommand.IValidateContainerCommandData context) {
+        try {
+            String dnsNamePrefix = context.getDnsNamePrefix();
+            Azure azureClient = context.getAzureClient();
+            String resourceGroupName = context.getResourceGroupName();
+
+            context.logStatus(
+                    String.format("Checking if the Azure Container Service with name 'containerservice-%s' exist.", dnsNamePrefix));
+            PagedList<ContainerService> containerServices = azureClient.containerServices().listByResourceGroup(resourceGroupName);
```
Comment (ArieShout) [inline src/main/java/com/microsoft/jenkins/acs/commands/ValidateContainerCommand.java]: This is a direct translation from the previous logic. Agreed we should use the concrete method. This class will be removed entirely  in later update of the plugin.

I will update the logic anyway to reduce confusion.

File: src/main/java/com/microsoft/jenkins/acs/util/NetworkResourceProviderHelper.java [hunk 7]
```diff
@@ -5,200 +5,162 @@
...
-		ipAddresses = ipAddressResponse.getPublicIpAddresses();
-		if(ipAddresses.size() != 2) {
-			throw new AzureCloudException("Not able to find FQDN for management public IP address.");
-		}
-
-		int index = ipAddresses.get(0).getDnsSettings().getFqdn().contains("@" + dnsNamePrefix + "mgmt.") ? 0 : 1;
-		return ipAddresses.get(index).getDnsSettings().getFqdn();
-	}
-	
+    public static String getMgmtPublicIPFQDN(Azure azureClient, String resourceGroupName, String dnsNamePrefix)
```
Comment (ArieShout) [inline src/main/java/com/microsoft/jenkins/acs/util/NetworkResourceProviderHelper.java]: It's currently still used in `GetPublicFQDNCommand`, however, it will be removed in later updates.

I will keep it in this change list.

File: src/main/java/com/microsoft/jenkins/acs/util/NetworkResourceProviderHelper.java (hunk 4 above)
Comment (ArieShout) [inline src/main/java/com/microsoft/jenkins/acs/util/NetworkResourceProviderHelper.java]: Nice catch. Updated to `fromAnyAddress()`.

File: src/main/java/com/microsoft/jenkins/acs/util/NetworkResourceProviderHelper.java (hunk 3 above)
Comment (ArieShout) [inline src/main/java/com/microsoft/jenkins/acs/util/NetworkResourceProviderHelper.java]: Checked in ACS, the default probes created use TCP protocol. Updated this to TCP.

File: pom.xml (hunk 1 above)
Comment (ArieShout) [inline pom.xml]: Updated to `0.2.0-SNAPSHOT`

File: pom.xml (hunk 2 above)
Comment (ArieShout) [inline pom.xml]: Added.

File: src/main/java/com/microsoft/jenkins/acs/ACSDeploymentContext.java (hunk 5 above)
Comment (ArchangelSDY) [inline src/main/java/com/microsoft/jenkins/acs/ACSDeploymentContext.java]: I think in most cases the configuration file may be checked out from source control or copied from upstream job. Maybe we should let user specify a location relative to the job's work space rather than an absolute path? Then we can assume the file is always on the machine where the job runs and do not need to worry about the network location.

File: src/main/java/com/microsoft/jenkins/acs/util/NetworkResourceProviderHelper.java (hunk 4 above)
Comment (ArieShout) [inline src/main/java/com/microsoft/jenkins/acs/util/NetworkResourceProviderHelper.java]: It seems there's some change in the SecurityRule API change in the SDK. In the portal we can define rules for some source network with some tag, e.g., **Internet**. And that's how it was implemented in the previous version of this plugin.

However, this is not available in the new SDK.

Comment (ArchangelSDY) [inline src/main/java/com/microsoft/jenkins/acs/util/NetworkResourceProviderHelper.java]: I looked at the inner implementation. `Internet` should be OK:

```java
    /**
     * The CIDR or source IP range. Asterix '*' can also be used to match all
     * source IPs. Default tags such as 'VirtualNetwork', 'AzureLoadBalancer'
     * and 'Internet' can also be used. If this is an ingress rule, specifies
     * where network traffic originates from.
     */
    @JsonProperty(value = "properties.sourceAddressPrefix", required = true)
    private String sourceAddressPrefix;
```

File: pom.xml [hunk 8]
```diff
@@ -1,141 +1,143 @@
...
+        <java.level>7</java.level>
+        <!-- Jenkins Test Harness version you use to test the plugin. -->
+        <!-- For Jenkins version >= 1.580.1 use JTH 2.x or higher. -->
+        <jenkins-test-harness.version>2.1</jenkins-test-harness.version>
+        <!-- Other properties you may want to use:
+             ~ hpi-plugin.version: The HPI Maven Plugin version used by the plugin..
+             ~ stapler-plugin.version: The Stapler Maven plugin version required by the plugin.
+        -->
+        <azuresdk.version>1.1.0</azuresdk.version>
+        <azure-credentials-plugin.version>1.1</azure-credentials-plugin.version>
```
Comment (chenkennt) [inline pom.xml]: 1.2

File: src/main/java/com/microsoft/jenkins/acs/ACSDeploymentContextDescriptor.java [hunk 9]
```diff
@@ -128,72 +113,60 @@ public ListBoxModel doFillLocationItems() {
         return model;
     }
 
-	public ListBoxModel doFillAgentVMSizeItems() {
+    public ListBoxModel doFillAgentVMSizeItems() {
         ListBoxModel model = new ListBoxModel();
```
Comment (chenkennt) [inline src/main/java/com/microsoft/jenkins/acs/ACSDeploymentContextDescriptor.java]: Change to read VM sizes from SDK.

Comment (chenkennt) [inline src/main/java/com/microsoft/jenkins/acs/ACSDeploymentContextDescriptor.java]: This PR is too big and mainly code refactor, you can do it in another PR.

File: src/main/java/com/microsoft/jenkins/acs/commands/MarathonDeploymentCommand.java [hunk 10]
```diff
@@ -5,103 +5,125 @@
...
+
+                if (execChnl.isClosed()) {
+                    if (in.available() > 0) {
+                        continue;
+                    }
+                    if (execChnl.getExitStatus() < 0) {
+                        throw new AzureCloudException("Error building or running docker image. Process exected with status: " +
+                                execChnl.getExitStatus());
+                    }
+                    System.out.println("exit-status: " + execChnl.getExitStatus());
```
Comment (chenkennt) [inline src/main/java/com/microsoft/jenkins/acs/commands/MarathonDeploymentCommand.java]: Use context.logStatus

File: src/main/java/com/microsoft/jenkins/acs/services/AzureManagementServiceDelegate.java [hunk 11]
```diff
@@ -6,203 +6,164 @@
...
             public String call() throws Exception {
-                ServiceDelegateHelper.getStorageManagementClient(config).getStorageAccountsOperations().
-                        checkNameAvailability("CI_SYSTEM");
+                AzureCredentials.ServicePrincipal servicePrincipal = AzureCredentials.getServicePrincipal(azureCredentialsId);
+                if (StringUtils.isBlank(servicePrincipal.getClientId())) {
+                    return "Failure: Cannot get the service principal from Azure credentials ID " + azureCredentialsId;
+                }
+
+                Azure azureClient = Azure.authenticate(DependencyMigration.buildAzureTokenCredentials(servicePrincipal)).withSubscription(servicePrincipal.getSubscriptionId());
+                azureClient.storageAccounts().checkNameAvailability("CI_SYSTEM");
```
Comment (chenkennt) [inline src/main/java/com/microsoft/jenkins/acs/services/AzureManagementServiceDelegate.java]: why do we check availability of CI_SYSTEM?

File: src/main/java/com/microsoft/jenkins/acs/services/AzureManagementServiceDelegate.java [hunk 12]
```diff
@@ -6,203 +6,164 @@
...
+                    case CANCELED:
+                    case FAILED:
+                    case DELETING:
+                        LOGGER.log(Level.INFO, "Failed({0}): {1}:{2}", new Object[]{state, type, resource});
+                        baseCommandData.logError(String.format("Failed(%s): %s:%s", state, type, resource));
+                        return false;
+                    case SUCCEEDED:
+                        baseCommandData.logStatus(
+                                String.format("Succeeded(%s): %s:%s", state, type, resource));
+                        completed--;
```
Comment (chenkennt) [inline src/main/java/com/microsoft/jenkins/acs/services/AzureManagementServiceDelegate.java]: break

File: src/main/java/com/microsoft/jenkins/acs/ACSDeploymentContext.java (hunk 5 above)
Comment (ArieShout) [inline src/main/java/com/microsoft/jenkins/acs/ACSDeploymentContext.java]: Yes, that should be the most common case. I will remove this part of copying logic.

File: pom.xml (hunk 8 above)
Comment (ArieShout) [inline pom.xml]: Oh, 1.2 was released. I will update the dependency.

File: src/main/java/com/microsoft/jenkins/acs/ACSDeploymentContextDescriptor.java (hunk 9 above)
Comment (ArieShout) [inline src/main/java/com/microsoft/jenkins/acs/ACSDeploymentContextDescriptor.java]: Yes, indeed. This PR is generally a migration of the original logic to the new SDK environment. I didn't alter the logic too much so that it may be useful (or maybe useless) when we need to reference the original logic in future.

In my local shadow, these logic had been removed as we should not have cluster creating logic in CI system.

File: src/main/java/com/microsoft/jenkins/acs/services/AzureManagementServiceDelegate.java (hunk 11 above)
Comment (ArieShout) [inline src/main/java/com/microsoft/jenkins/acs/services/AzureManagementServiceDelegate.java]: I have no idea. This is a translation from the original code logic. I interpreted this as testing if we can actually connect to the Azure environment using the filled credentials, via a no-op operation.

In the latest code, this was removed as we will check that in the `azure-credentials` plugin.

File: src/main/java/com/microsoft/jenkins/acs/services/AzureManagementServiceDelegate.java (hunk 12 above)
Comment (ArieShout) [inline src/main/java/com/microsoft/jenkins/acs/services/AzureManagementServiceDelegate.java]: added.

File: src/main/java/com/microsoft/jenkins/acs/commands/MarathonDeploymentCommand.java (hunk 10 above)
Comment (ArieShout) [inline src/main/java/com/microsoft/jenkins/acs/commands/MarathonDeploymentCommand.java]: fixed.
